### PR TITLE
[InteractionRegions] Generated clip paths are sometimes offset from their element

### PR DESCRIPTION
**Pull request by etiennesegonzac**
#### 2775f5bf7c3e8aeb9b44090c7a277aedcdb7d0e4
<pre>
[InteractionRegions] Generated clip paths are sometimes offset from their element
<a href="https://bugs.webkit.org/show_bug.cgi?id=273428">https://bugs.webkit.org/show_bug.cgi?id=273428</a>
&lt;<a href="https://rdar.apple.com/126865156">rdar://126865156</a>&gt;

Reviewed by Simon Fraser.

When generating Interaction Regions, the paths we build to match
shapes are complete and relative to the element&apos;s bounds.
But the Interaction Region&apos;s bounds respect clipping.
Provide the offset between the two to `interactionRegionForRenderedRegion`
so we can adjust the paths location if needed.
(Effectively moving the path in the mask layer on the UI-side.)

Also provide the current transform to skip shape generation for
rotations / skew (unsupported) and also to  avoid having to go through
`absoluteQuads` just to get a bounding size.

This applies to SVG elements, icons and CSS clip-paths.

For SVG icons (using image tags), always use the image from the
SVGImageCache to match what was just painted.

* Source/WebCore/page/InteractionRegion.h:
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::boundingSize):
(WebCore::absoluteBoundingRect): Deleted.
Don&apos;t use `absoluteQuads` to get the bounding size.
(WebCore::findIconImage):
Use `imageForRenderer` to make sure icon masking is based on what was
just painted (and potentially cached in the SVGImageCache).
(WebCore::interactionRegionForRenderedRegion):
For icons, generate the raster shape at 0x0 directly instead of doing an
extra translation.
For SVG elements, use the full viewBoxTransform instead of just the
scale.
Use the new `clipOffset`.

* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::unite):
(WebCore::EventRegionContext::uniteInteractionRegions):
Pass an optional transform and the clip offset to
`interactionRegionForRenderedRegion`.

* LayoutTests/interaction-region/icon-masking.html:
Add test cases clipping on the top left corner (bitmap and SVG).
Add test cases with nested transforms (with and without compositing
layers).
Add a test case with rotation (no clip-path).
* LayoutTests/interaction-region/icon-masking-expected.txt:
New interaction regions, new guards on the SVG logo now matching the &quot;is
complex shape&quot; check, and floating point diff.
* LayoutTests/interaction-region/svg.html:
Add a test case clipping on the top left corner.
* LayoutTests/interaction-region/svg-expected.txt:
New interaction regions and floating point diff.
* LayoutTests/interaction-region/clip-path-expected.txt:
Add a test case clipping on the left and right edge.
Add test cases with rotation / skew (no clip-path).
* LayoutTests/interaction-region/clip-path.html:
New interaction regions.

Canonical link: <a href="https://commits.webkit.org/278371@main">https://commits.webkit.org/278371@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/d006798aafa2e74576e9fdcecd01a39ac5ad0695

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50243 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29536 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2540 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53502 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/933 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52546 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35763 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/564 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40999 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52342 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27204 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43249 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22098 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24622 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/501 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8621 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46608 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/563 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55087 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25339 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/493 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48399 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26600 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43433 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47422 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11043 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27464 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26332 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/page/InteractionRegion.cpp [hunk 1]
```diff
@@ -444,34 +458,34 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
         auto boundingRect = absoluteBoundingRect(regionRenderer);
         clipPath = styleClipPath->getPath(TransformOperationData(FloatRect(FloatPoint(), boundingRect.size())));
     } else if (iconImage && originalElement) {
-        LayoutRect imageRect(rect);
+        auto boundingRect = absoluteBoundingRect(regionRenderer);
```
Comment (smfr) [inline Source/WebCore/page/InteractionRegion.cpp]: Calling `absoluteBoundingRect` just to get its size is wasteful; it pays the cost for mapping the rect to absolute coordinates.

Comment (etiennesegonzac) [reply Source/WebCore/page/InteractionRegion.cpp]: ~~we use the location in `pathTranslationToMatchClipping` below~~

no we don't, updating :)

Comment (etiennesegonzac) [reply Source/WebCore/page/InteractionRegion.cpp]: we still need the transform but this should be better

File: Source/WebCore/page/InteractionRegion.cpp [hunk 2]
```diff
@@ -441,37 +463,37 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
     RefPtr styleClipPath = regionRenderer.style().clipPath();
 
     if (styleClipPath && styleClipPath->type() == PathOperation::OperationType::Shape && originalElement) {
-        auto boundingRect = absoluteBoundingRect(regionRenderer);
-        clipPath = styleClipPath->getPath(TransformOperationData(FloatRect(FloatPoint(), boundingRect.size())));
+        auto size = boundingSize(regionRenderer, transform);
```
Comment (smfr) [inline Source/WebCore/page/InteractionRegion.cpp:457]: I don't know if this transform thing is correct. You've deviated so far from the painting code that it's impossible to know how much of this transform/clipping logic is OK.

File: Source/WebCore/page/InteractionRegion.cpp [hunk 3]
```diff
@@ -279,6 +287,20 @@ static std::optional<std::pair<Ref<SVGSVGElement>, Ref<SVGGraphicsElement>>> fin
     return std::nullopt;
 }
 
+static FloatSize pathTranslationToMatchClipping(const FloatRect& clippedRect, const std::optional<FloatRect>& clippingRect, const FloatSize& boundingSize)
+{
+    if (!clippingRect)
+        return FloatSize();
+
+    auto clipDelta = FloatSize(clippedRect.width() - boundingSize.width(), clippedRect.height() - boundingSize.height());
```
Comment (smfr) [inline Source/WebCore/page/InteractionRegion.cpp]: This seems really ad-hoc. What is it doing?

File: Source/WebCore/rendering/EventRegion.cpp [hunk 4]
```diff
@@ -71,7 +71,15 @@ void EventRegionContext::unite(const FloatRoundedRect& roundedRect, RenderObject
...
         return affineTransform.mapRect(rect);
     });
-    uniteInteractionRegions(renderer, layerBounds);
+
+    std::optional<FloatRect> clipRect;
+    if (!m_clipStack.isEmpty())
+        clipRect = m_clipStack.last();
+    std::optional<AffineTransform> transform;
+    if (!m_transformStack.isEmpty())
+        transform = m_transformStack.last();
```
Comment (smfr) [inline Source/WebCore/rendering/EventRegion.cpp:78]: Is just grabbing the last transform on the stack OK? What transform do you need; a root-relative one, or one relative to some ancestor? Which ancestor?

Comment (etiennesegonzac) [reply Source/WebCore/rendering/EventRegion.cpp:78]: We want to apply the same transform / clipping as the one applied to the layer bounds in `transformAndClipIfNeeded`. (it only uses the last clip rect / transform)

I'm adding a couple of tests with nested transforms

File: Source/WebCore/page/InteractionRegion.cpp [hunk 5]
```diff
@@ -226,11 +226,19 @@ static bool isGuardContainer(const Element& element)
...
-    return unitedBoundingBoxes(quads);
+    Vector<LayoutRect> rects;
+    renderer.boundingRects(rects, LayoutPoint());
+
+    if (!rects.size())
+        return FloatSize();
+
+    FloatSize size = unionRect(rects).size();
+    if (transform)
+        size.scale(transform->xScale(), transform->yScale());
```
Comment (smfr) [inline Source/WebCore/page/InteractionRegion.cpp:239]: Why only scale? What if the transform is a skew or rotate?

**Review by smfr — APPROVED** [review]

**Pull request merged**